### ISSUE TITLE
Inefficient(?) mapping of Haskell String/Text types to "char *"

### ISSUE THREAD

**Issue by clojj** [tree-sitter/haskell-tree-sitter#79]
There are several language-bindings (not grammar but the TS API bindings!) around, and I guess that this issue may not be as obvious or performance-relevant in these other bindings.
But in Haskell, I am a bit uncomfortable with this allocation, everytime I need to call TS with some input string (source to be parsed):
```
(str, len) <- newCStringLen source
tree <- ts_parser_parse_string parser nullPtr str len
```
...and when you (for good reasons) have `Data.Text` instead of `String`, it is even necessary to even `unpack` that first before the newCStringLen-allocation.

Given that the real performance-hit of this is yet to be evaluated, would it be a valid line of thought, to have some "callback mechanism" on the tree-sitter side, so that each time tree-sitter and it's scanner needs some characters for lexing/parsing, it can get these by such a callback function ?
I can imagine a Haskell FFI callback (Haskell function callable from C), which then provides chars as needed during lexing/parsing...

Not trivial, but is this thought more on the crazy side or more on the "theoretically do-able" side  ?

Regards

**Comment by clojj** [comment]
when I think about it.... maybe this is an issue for tree-sitter, not here (?)

**Comment by robrix** [comment]
We handle this via `unsafeUseAsCStringLen` & `bracket`, something like so:

```haskell
unsafeUseAsCStringLen byteString $ \ (bytes, len) -> do
  let acquire = ts_parser_parse_string parser nullPtr bytes len
      release t
        | t == nullPtr = pure ()
        | otherwise = ts_tree_delete t
  bracket acquire release (\ treePtr -> …)
```

**Comment by clojj** [comment]
great.. didn’t see that possibility, I will try this then
closing the issue, thanks!

**Comment by maxbrunsfeld** [comment]
There already is a callback mechanism. The function is called ‘ts_parser_parse’. It takes a function pointer instead of a C string.

Tree-sitter can also parse UTF16 strings directly, which might be useful as I believe Haskell’s Text type uses UTF16 internally. I’m not sure if Text gives you a way to get a raw pointer to the UTF16 data though.

**Comment by maxbrunsfeld** [comment]
Oh, @robrix beat me to this with a clearer answer! ✨

**Comment by clojj** [comment]
yippee, thx

**Comment by clojj** [comment]
> Oh, @robrix beat me to this with a clearer answer! ✨

your suggestion is very interesting as well !

**Comment by clojj** [comment]
@maxbrunsfeld 
you mean this function ?
https://github.com/tree-sitter/tree-sitter/blob/6d97cbd2b6719c016a23f05ce874fba760ecae84/include/tree_sitter/runtime.h#L88

**Comment by maxbrunsfeld** [comment]
Yup, exactly. [`TSInput`](https://github.com/tree-sitter/tree-sitter/blob/6d97cbd2b6719c016a23f05ce874fba760ecae84/include/tree_sitter/runtime.h#L43-L47) is a `read` callback plus a `void *` arbitrary data pointer.

**Comment by clojj** [comment]
both of your answers pointed me to efficient Data.Text handling.. utf16 encoding works as expected so far
I will probably have to adjust the byte offsets of the parsed nodes now, and that should be it
After verifying these byteoffsets, I will close this issue.

**Comment by clojj** [comment]
as far as actual performance goes, unfortunately I don't have a benchmark harness yet, but it's good to know this works as expected anyway.
I'll leave it for now, because that'd be premature optimization :smirk:
